server/src/server.js

```diff
@@ -189,6 +215,76 @@ app.get("/:id/:domain", function (req, res) {
   });
 });
 
+// Get OAuth client params for the client-side authorization flow.
+app.get('/api/fxa-oauth/params', function (req, res, next) {
+  if (! req.userId) {
+    next(errors.missingSession());
+    return;
+  }
+  randomBytes(32).then(stateBytes => {
+    let state = stateBytes.toString('hex');
+    return setState(req.userId, state).then(inserted => {
+      if (!inserted) {
+        throw errors.dupeLogin();
+      }
+      return state;
+    });
+  }).then(state => {
+    res.send({
+      // FxA profile server URL.
+      profile_uri: config.oAuth.profileServer,
+      // FxA OAuth server URL.
+      oauth_uri: config.oAuth.oAuthServer,
+      redirect_uri: 'urn:ietf:wg:oauth:2.0:fx:webchannel',
+      client_id: config.oAuth.clientId,
+      // FxA content server URL.
+      content_uri: config.oAuth.contentServer,
+      state,
+      scope: 'profile'
+    });
+  }).catch(next);
+});
+
+// Exchange an OAuth authorization code for an access token.
+app.post('/api/fxa-oauth/token', function (req, res, next) {
+  if (! req.userId) {
+    next(errors.missingSession());
+    return;
+  }
+  if (! req.body) {
+    next(errors.missingParams());
+    return;
+  }
+  let { code, state } = req.body;
+  checkState(req.userId, state).then(isValid => {
+    if (!isValid) {
+      throw errors.badState();
+    }
+    return tradeCode(code);
+  }).then(({ access_token: accessToken }) => {
+    return getAccountId(accessToken).then(({ uid: accountId }) => {
+      return registerAccount(req.userId, accountId, accessToken);
+    }).then(() => {
+      res.send({
+        access_token: accessToken
+      });
+    });
+  }).catch(next);
+});
+
+app.use(function (err, req, res, next) {
```

Comment:
I moved the error handler last so that `.catch(next)` can catch and report errors in promise chains.